go-controller/pkg/ovn/gateway.go

```diff
@@ -403,7 +403,7 @@ func (gw *GatewayManager) GatewayInit(
 			types.NetworkExternalID:  gw.netInfo.GetNetworkName(),
 			types.TopologyExternalID: gw.netInfo.TopologyType(),
 		}
-		if gw.netInfo.IsPrimaryNetwork() && gw.netInfo.TopologyType() == types.Layer2Topology {
+		if gw.netInfo.IsPrimaryNetwork() && gw.netInfo.TopologyType() == types.Layer2Topology && config.IPv6Mode {
```

Comment:
lgtm

I pray some unit test somewhere screams, I'll be sad otherwise.... cause that means we are not testing single stack also in UDN UTs...

Reply:
I spoke too soon! UTs failing, its checked in a different place:
```
2024-10-28T20:47:32.0392453Z [38;5;9m[1mSummarizing 4 Failures:[0m
2024-10-28T20:47:32.0394836Z   [38;5;9m[FAIL][0m [0mOVN Multi-Homed pod operations for layer2 network [38;5;243mreconciles a new [38;5;9m[1m[It] pod on a user defined primary network[0m
2024-10-28T20:47:32.0397915Z   [38;5;243m/home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/pkg/ovn/secondary_layer2_network_controller_test.go:161[0m
2024-10-28T20:47:32.0400471Z   [38;5;9m[FAIL][0m [0mOVN Multi-Homed pod operations for layer2 network [38;5;243mreconciles a new [38;5;9m[1m[It] pod on a user defined primary network on an IC cluster[0m
2024-10-28T20:47:32.0402407Z   [38;5;243m/home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/pkg/ovn/secondary_layer2_network_controller_test.go:161[0m
2024-10-28T20:47:32.0403952Z   [38;5;9m[FAIL][0m [0mOVN Multi-Homed pod operations for layer2 network [38;5;243mreconciles a new [38;5;9m[1m[It] pod on a user defined primary network on an IC cluster; LGW[0m
2024-10-28T20:47:32.0405404Z   [38;5;243m/home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/pkg/ovn/secondary_layer2_network_controller_test.go:161[0m
2024-10-28T20:47:32.0407011Z   [38;5;9m[FAIL][0m [0mOVN Multi-Homed pod operations for layer2 network [38;5;243mreconciles a new [38;5;9m[1m[It] pod on a user defined primary network on an IC cluster with per-pod SNATs enabled[0m
2024-10-28T20:47:32.0408536Z   [38;5;243m/home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/pkg/ovn/secondary_layer2_network_controller_test.go:161[0m
```

```
2024-10-28T20:41:44.4461783Z               expected: <*nbdb.LogicalRouterPort | 0xc03030dc80>{
2024-10-28T20:41:44.4462500Z                   UUID: "rtos-isolatednet_ovn_layer2_switch-UUID",
2024-10-28T20:41:44.4462974Z                   DhcpRelay: nil,
2024-10-28T20:41:44.4463334Z                   Enabled: nil,
2024-10-28T20:41:44.4463699Z                   ExternalIDs: {
2024-10-28T20:41:44.4464218Z                       "k8s.ovn.org/topology": "layer2",
2024-10-28T20:41:44.4464818Z                       "k8s.ovn.org/network": "isolatednet",
2024-10-28T20:41:44.4465201Z                   },
2024-10-28T20:41:44.4465408Z                   GatewayChassis: nil,
2024-10-28T20:41:44.4465616Z                   HaChassisGroup: nil,
2024-10-28T20:41:44.4465905Z                   Ipv6Prefix: nil,
2024-10-28T20:41:44.4466095Z                   Ipv6RaConfigs: {
2024-10-28T20:41:44.4466426Z                       "address_mode": "dhcpv6_stateful",
2024-10-28T20:41:44.4466625Z                       "mtu": "1400",
2024-10-28T20:41:44.4466964Z                       "send_periodic": "true",
2024-10-28T20:41:44.4467216Z                       "max_interval": "900",
2024-10-28T20:41:44.4467462Z                       "min_interval": "300",
2024-10-28T20:41:44.4467761Z                       "router_preference": "LOW",
2024-10-28T20:41:44.4467992Z                   },
```

Reply:
The test cases should pass now. There is no coverage for ipv6. I tried to add it but everything is hardcoded to ipv4 in these tests. Its going to take some time (which I dont have) to fix it. I added a FIXME and hopefully someone else can go fix the tests. For now at least the tests work correctly with IPv4.

Reply:
yeah no problem do you want to open an upstream issue to track the missing v6 tests maybe @maiqueb has some time since he knows these tests the best

Reply:
I'll try to improve IPv6 coverage in the following days.